src/meta/src/barrier/mod.rs

```diff
@@ -785,55 +839,54 @@ where
-                        .cloned()
-                        .map(|grouped| {
-                            let sst = grouped.sst.expect("field not None");
-                            sst_to_worker.insert(sst.id, resp.worker_id);
-                            (grouped.compaction_group_id, sst)
-                        })
-                        .collect_vec();
-                    synced_ssts.append(&mut t);
-                }
-
-                if prev_epoch == INVALID_EPOCH {
-                    assert!(
-                        synced_ssts.is_empty(),
-                        "no sstables should be produced in the first epoch"
-                    );
-                } else {
-                    self.hummock_manager
-                        .commit_epoch(prev_epoch, synced_ssts, sst_to_worker)
-                        .await?;
-                }
-
                 node.timer.take().unwrap().observe_duration();
-                node.wait_commit_timer.take().unwrap().observe_duration();
-                node.command_ctx.post_collect().await?;
 
-                // Notify about collected first.
-                let mut notifiers = take(&mut node.notifiers);
-                notifiers.iter_mut().for_each(Notifier::notify_collected);
-
-                // Then try to finish the barrier for Create MVs.
-                let actors_to_finish = node.command_ctx.actors_to_track();
-                tracker.add(node.command_ctx.curr_epoch, actors_to_finish, notifiers);
-                for progress in resps.iter().flat_map(|r| &r.create_mview_progress) {
-                    tracker.update(progress);
+                let notifiers = take(&mut node.notifiers);
+                let command_ctx = node.command_ctx.clone();
+                let create_mv_progress = resps
+                    .iter()
+                    .flat_map(|r| r.create_mview_progress.clone())
+                    .collect_vec();
+                checkpoint_control
+                    .add_uncommitted_states(resps, (command_ctx, notifiers, create_mv_progress));
+
+                // If not sync , we can't notify collection completion
+                if resps.iter().all(|node| node.need_sync) {
+                    let mut uncommitted_states = checkpoint_control.get_uncommitted_states();
+                    if prev_epoch != INVALID_EPOCH {
+                        self.hummock_manager
+                            .commit_epoch(
+                                prev_epoch,
+                                uncommitted_states.uncommitted_ssts,
+                                uncommitted_states.uncommitted_work_id,
+                            )
+                            .await?;
+                    }
+                    while let Some((command_ctx, mut notifiers, create_mv_progress)) =
+                        uncommitted_states.uncommitted_checkpoint_post.pop_back()
+                    {
+                        checkpoint_control.remove_changes(command_ctx.command.changes());
```

Comment:
Prior to this PR, `remove_changes` happens before `commit_epoch` while this PR makes `commit_epoch` happen first. Will this be a problem?

Reply:
I think it is not a problem. We just need to make sure that both `remove_changes` and `update actor msg` happen before the next `inject_barrier`